### DISCUSSION TITLE
Can Serilog log messages be redirected to Standard Out Messages in xunit.v3 ?

### DISCUSSION THREAD

**Discussion by PureKrome** [xunit/xunit#3081] [Question] [answered]
👋🏻 Hi again team!

I'm testing my asp.net 8 webapi program/app.
In it, I use Serilog to handle logging. Works great!

Now in my app during startup I have some bad code (related to EFCore design/relationships). It throws an exception. My program.cs has a 'main' try/catch and in the catch we (serilog) Log the error message.

In my test, the code fails (as expected) but I can't see the logging at all anywhere. Is there any way I can see this logging so I know that the program.cs failed and what the failure is? Currently, I need to "step in" the code to see it thrown and read the exception in the try/catch.

here's the program.cs code, to give you an example:

```
Log.Logger = new LoggerConfiguration()
    .WriteTo.Console()
    .CreateBootstrapLogger();

try
{
    var builder = WebApplication.CreateBuilder(args);

    // SNIP SNIP

    var app = builder.Build();

    app.Logger.LogInformation("Starting application");

    // ----------------------------------------------------------------------------

   // SNIP SNIP, again

    await app.RunAsync();
}
catch (Exception exception)
{
    Log.Fatal(exception, "Unhandled startup exception.");
    Log.Fatal("Exception: {Exception}", exception);
}
finally
{
    await Log.CloseAndFlushAsync();
}

// This enables testing to work.
public partial class Program
{
}
```

and then in some test code, we use the WebApplicationFactory:

```
public sealed class CustomWebApplicationFactory(string connectionString) : WebApplicationFactory<Program>
{
    protected override IHost CreateHost(IHostBuilder builder)
    {
        // SNIP
    }

    protected override void ConfigureWebHost(IWebHostBuilder builder)
    {
        // SNIP
    }
}
```

So is it possible to update Serilog to goto Standard Output so the test output can display any/all messages?

I tried creating my own sink:

```
public class TestOutputHelperSink : ILogEventSink
{
    private readonly ITestOutputHelper _testOutputHelper;

    public TestOutputHelperSink(ITestOutputHelper testOutputHelper)
    {
        _testOutputHelper = testOutputHelper;
    }

    public void Emit(LogEvent logEvent)
    {
        _testOutputHelper.WriteLine(logEvent.RenderMessage());
    }
}
```

and then (like in a test or somewhere):

```
        Log.Logger = new LoggerConfiguration()
            .MinimumLevel.Debug()
            .WriteTo.Console()
            .WriteTo.Sink(new TestOutputHelperSink(_testOutputHelper))
            .CreateLogger();
```

Any help would be so appreciated!

**Comment by bradwilson** [comment, marked as answer]
You can capture console output as test output with:

```csharp
[assembly: Xunit.CaptureConsole]
```

All console output will show up as test output (akin to when you write to test output with `ITestOutputHelper`).

**Reply by PureKrome** [reply]
Winner - this works. Legend, @bradwilson 💪🏻 

Thank you for your time and help!
